File: scsc-monolith/src/main/java/demo/scsc/commandside/payment/OrderPayment.java
package demo.scsc.commandside.payment;

import demo.scsc.api.payment.*;
import org.axonframework.commandhandling.CommandExecutionException;
import org.axonframework.commandhandling.CommandHandler;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.eventhandling.EventMessage;
import org.axonframework.eventsourcing.EventSourcingHandler;
import org.axonframework.messaging.InterceptorChain;
import org.axonframework.messaging.interceptors.MessageHandlerInterceptor;
import org.axonframework.modelling.command.AggregateIdentifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.util.UUID;

import static org.axonframework.modelling.command.AggregateLifecycle.*;

public class OrderPayment {

    private static final Logger LOG = LoggerFactory.getLogger(OrderPayment.class);

    @AggregateIdentifier
    private UUID orderPaymentId;
    private UUID orderId;
    private BigDecimal requestedAmount;
    private BigDecimal paidAmount = BigDecimal.ZERO;

    @SuppressWarnings("unused")
    public OrderPayment() {
        // required no-arg constructor
    }

    @CommandHandler
    public OrderPayment(RequestPaymentCommand requestPaymentCommand) {

        /* -------------------------
                validation
        ------------------------- */

        if (requestPaymentCommand.orderId() == null) {
            throw new CommandExecutionException("Can't request payment without order", null);
        }

        if (requestPaymentCommand.amount().compareTo(BigDecimal.ZERO) <= 0) {
            throw new CommandExecutionException("Can't process payments for zero or negative amounts", null);
        }

        /* -------------------------
                notification
        ------------------------- */

        apply(new PaymentRequestedEvent(
                requestPaymentCommand.orderPaymentId(),
                requestPaymentCommand.orderId(),
                requestPaymentCommand.amount()
        ));
    }

    @CommandHandler
    public void on(ProcessPaymentCommand processPaymentCommand) {
        BigDecimal leftToPay = requestedAmount.subtract(paidAmount);

        if (processPaymentCommand.amount().compareTo(leftToPay) > 0) {
            throw new CommandExecutionException("Can't pay more than you own", null);
        }

        apply(new PaymentReceivedEvent(orderPaymentId, processPaymentCommand.amount()));

        if (processPaymentCommand.amount().compareTo(leftToPay) == 0) {
//            apply(new OrderFullyPaidEvent(orderId, orderPaymentId));
            apply(new OrderFullyPaidEvent(orderPaymentId, orderId));

        }
    }

    @EventSourcingHandler
    public void on(PaymentRequestedEvent paymentRequestedEvent) {
        this.orderPaymentId = paymentRequestedEvent.orderPaymentId();
        this.orderId = paymentRequestedEvent.orderId();
        this.requestedAmount = paymentRequestedEvent.amount();
    }

    @EventSourcingHandler
    public void on(PaymentReceivedEvent paymentReceivedEvent) {
        this.paidAmount = this.paidAmount.add(paymentReceivedEvent.amount());
    }

    @EventSourcingHandler
    public void on(OrderFullyPaidEvent orderFullyPaidEvent) {
        markDeleted();
    }


    @MessageHandlerInterceptor(messageType = CommandMessage.class)
    public void intercept(CommandMessage<?> message,
                          InterceptorChain interceptorChain) throws Exception {
        LOG.info("[  COMMAND ] " + message.getPayload().toString());
        interceptorChain.proceed();
    }

    @MessageHandlerInterceptor(messageType = EventMessage.class)
    public void intercept(EventMessage<?> message,
                          InterceptorChain interceptorChain) throws Exception {
        if (isLive()) {
            LOG.info("[    EVENT ] " + message.getPayload().toString());
        } else {
            LOG.info("[ SOURCING ] " + message.getPayload().toString());
        }
        interceptorChain.proceed();
    }

}

File: scsc-monolith/src/main/kotlin/demo/scsc/commandside/payment/OrderPayment.kt
package demo.scsc.commandside.payment

import demo.scsc.api.payment
import demo.scsc.api.payment.RequestPaymentCommand
import org.axonframework.commandhandling.CommandExecutionException
import org.axonframework.commandhandling.CommandHandler
import org.axonframework.commandhandling.CommandMessage
import org.axonframework.eventhandling.EventMessage
import org.axonframework.eventsourcing.EventSourcingHandler
import org.axonframework.extensions.kotlin.applyEvent
import org.axonframework.messaging.InterceptorChain
import org.axonframework.messaging.interceptors.MessageHandlerInterceptor
import org.axonframework.modelling.command.AggregateIdentifier
import org.axonframework.modelling.command.AggregateLifecycle.isLive
import org.axonframework.modelling.command.AggregateLifecycle.markDeleted
import org.slf4j.LoggerFactory
import java.math.BigDecimal
import java.util.*

class OrderPayment() {
    @AggregateIdentifier
    internal lateinit var orderPaymentId: UUID
    internal lateinit var orderId: UUID
    internal lateinit var requestedAmount: BigDecimal
    internal var paidAmount = BigDecimal.ZERO

    @CommandHandler
    constructor(command: RequestPaymentCommand) : this() {
        if (command.amount <= BigDecimal.ZERO)
            throw CommandExecutionException("Can't process payments for zero or negative amounts", null)

        applyEvent(
            payment.PaymentRequestedEvent(
                command.orderPaymentId,
                command.orderId,
                command.amount
            )
        )
    }

    @CommandHandler
    fun on(command: payment.ProcessPaymentCommand) {
        val leftToPay = requestedAmount.subtract(paidAmount)
        if (command.amount > leftToPay) {
            throw CommandExecutionException("Can't pay more than you owe", null)
        }
        applyEvent(payment.PaymentReceivedEvent(orderPaymentId, command.amount))
        if (command.amount.compareTo(leftToPay) == 0) {
            applyEvent(payment.OrderFullyPaidEvent(orderPaymentId, orderId))
        }
    }

    @EventSourcingHandler
    fun on(event: payment.PaymentRequestedEvent) {
        orderPaymentId = event.orderPaymentId
        orderId = event.orderId
        requestedAmount = event.amount
    }

    @EventSourcingHandler
    fun on(event: payment.PaymentReceivedEvent) {
        paidAmount = paidAmount.add(event.amount)
    }

    @Suppress("UNUSED_PARAMETER")
    @EventSourcingHandler
    fun on(event: payment.OrderFullyPaidEvent) {
        markDeleted()
    }

    @MessageHandlerInterceptor(messageType = CommandMessage::class)
    fun intercept(message: CommandMessage<*>, interceptorChain: InterceptorChain) {
        LOG.info("[  COMMAND ] ${message.payload}")
        interceptorChain.proceed()
    }

    @MessageHandlerInterceptor(messageType = EventMessage::class)
    fun intercept(message: EventMessage<*>, interceptorChain: InterceptorChain) {
        if (isLive()) LOG.info("[    EVENT ] ${message.payload}") else LOG.info("[ SOURCING ] ${message.payload}")
        interceptorChain.proceed()
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(OrderPayment::class.java)
    }
}
